### ISSUE TITLE
Port Plurimath to JavaScript using Opal

### ISSUE THREAD

**Issue by hmdne** [plurimath/plurimath#159]
Opal is a source-to-source Ruby to JavaScript compiler and runtime. This issue is about exposing Plurimath to the JavaScript environments. Below is a list of items that will be updated over time to denote the status of the operation. This is a tracking bug that will be referenced by all upstreaming efforts. Not all tasks are necessary to completion. This list will be updated.

- [x] port to Oga (unlike Nokogiri or Ox, it is _almost_ a pure Ruby implementation, and Opal can only support Ruby and JavaScript for compilation - not C/Java extensions)
  - [x] make RSpec reach >50%
  - [x] make RSpec reach 100%
- [x] add Opal-RSpec
  - [x] make Opal-RSpec reach >50%
  - [ ] make Opal-RSpec reach 100%
- [x] generate a bundle
- [ ] upstream changes
  - [ ] sent
  - [ ] merged
  - [ ] released
- [x] fix a problem with updated Equivalent-XML
- [x] add JavaScript API
- [ ] generate NPM package
- [ ] port/fix hard dependencies
  - [x] Opal
    - [x] fix issues
    - [ ] upstream changes
      - [x] sent
      - [x] merged
      - [ ] released
  - [x] Oga
    - [x] port to pure-Ruby
      - [x] make RSpec reach >50%
      - [ ] make RSpec reach 100%
    - [x] add Opal-RSpec
      - [x] make Opal-RSpec reach >50%
      - [ ] make Opal-RSpec reach 100%
    - [ ] upstream changes
      - [ ] sent
      - [ ] merged
      - [ ] released
  - [x] Ruby-LL
    - [x] port to pure-Ruby
      - [x] make RSpec reach >50%
      - [x] make RSpec reach 100%
    - [x] add Opal-RSpec
      - [x] make Opal-RSpec reach >50%
      - [ ] make Opal-RSpec reach 100%
    - [ ] upstream changes
      - [x] sent
      - [ ] merged
      - [ ] released
  - [x] Parslet
    - [ ] investigate a performance issue
    - [x] add Opal-RSpec
      - [x] make Opal-RSpec reach >50%
      - [ ] make Opal-RSpec reach 100%
    - [ ] upstream changes
      - [ ] sent
      - [ ] merged
      - [ ] released
  - [x] htmlentities
    - [ ] upstream changes
      - [ ] sent
      - [ ] merged
      - [ ] released
- [ ] port soft dependencies (ie. needed for the specs to pass)
  - [ ] Equivalent-XML
    - [x] fix a compatibility problem
    - [ ] pass tests before submitting patches
    - [ ] upstream changes
      - [ ] sent
      - [ ] merged
      - [ ] released
  - [x] Flexmock
    - [x] fix a compatibility problem
    - [ ] upstream changes
      - [x] sent
      - [ ] merged
      - [ ] released

**Comment by hmdne** [comment]
Current state:

Finished in 11 minutes 49 seconds (files took 4.59 seconds to load)
2782 examples, 105 failures

The time improvement is related to the latest work on Opal, but Parslet performance issue still persists and will need to be investigated further.

I am preparing a set of repositories that will allow others to reproduce a workable state. This is far from fully finished, but unfortunately will require more work at upstreaming (if upstream parties will want to participate). Also, there are certain issues that still will need to be worked out before we can claim 100% test coverage.

I have created a GitHub organization to be transfered to you (or removed and integrated into this org) that contains all patched libraries (at the time of writing this, I'm still uploading the repositories). I have devised a main repo, that references submodules, which at this state, facilitates ease of development (the bug fixing process involves editing all repositories at once, but when upstreaming status is clarified, we will fork and publish relevant libraries and remove the submodules).

https://github.com/plurimath-js/plurimath-js
